src/MultiCollateralVault.sol

```diff
@@ -0,0 +1,346 @@
+ */
+
+    /// @return price the value of token denominated in WETH
+    /// @param token address for the token that's being inquired about
+    function price(address token) public view returns(uint256) {
+        return _price(token);
+    }
+
+    /// @notice calculates the total value of a users deposits in ETH
+    function totalDeposits() view external returns(uint256) {
+        return _totalDeposits(msg.sender);
+    }
+
+    /// @notice provides how much WETH a user is currently able to borrow
+    function borrowingPower(address user) view external returns(uint256) {
+        return _borrowingPower(user);
+    }
+
+    /// @notice provides a user's maximum combined loan to value ratio
+    function maxRatio() view external returns(uint256) {
+        return _maxRatio(msg.sender);
+    }
+
+    /// @notice provides a user's current loan to value ratio
+    function ltvRatio() view external returns(uint256) {
+        return _ltvRatio(msg.sender);
+    }
+
+    function freeCollateral(address user, address token) view external returns(uint256) {
+        return _freeCollateral(user, token);
+    }
+    
+    function _price(address token) internal view returns(uint256) {
+        AggregatorV3Interface iOracle = Oracles[token];
+        (   /*uint80 roundID*/,
+        int256 answer,
+            /*uint startedAt*/,
+            /*uint timeStam*/,
+            /*uint80 answeredInRound*/
+        ) = iOracle.latestRoundData();
+        return uint256(answer);
+    }
+
+    /// @dev calculates the the ratioed value of a users deposits in ETH
+    function _ratioedDeposits(address user) internal view returns(uint256) {
+        uint256 ratioedValue;
+        address[] memory _collaterals = Collaterals;
+        uint256 _length = _collaterals.length;
+        for(uint256 i = 0; i < _length; ) {
+            ratioedValue += RMul.rmul(
+                WMul.wmul(
+                Deposits[user][_collaterals[i]]*10**_decimalsDifference(_collaterals[i]),
+                _price(_collaterals[i])), 
+                Ratios[_collaterals[i]]);
+                unchecked{i++;}
+        }
+        return ratioedValue;
+    }
+
+    
+    function _totalDeposits(address user) internal view returns(uint256) {
+        uint256 totalValue;
+        address[] memory _collaterals = Collaterals;
+        uint256 _length = _collaterals.length;
+        for(uint256 i = 0; i < _length;){
+            totalValue += WMul.wmul(Deposits[user][_collaterals[i]]*10**_decimalsDifference(_collaterals[i]), _price(_collaterals[i]));
+            unchecked{i++;}
+        }
+        return totalValue;
+    }
+
+    function _borrowingPower(address user) internal view returns(uint256) {
+        return _ratioedDeposits(user) - Debt[user];
+    }
+
+    /// @notice calculates the maximum LTV ratio of a users vault
+    function _maxRatio(address user) internal view returns(uint256){
+        return WDiv.wdiv(_ratioedDeposits(user), _totalDeposits(user));
```

Comment:
Both `_ratioedDeposits` and `_totalDeposits` check the price of each collateral from the oracle, which is quite expensive. It would be much cheaper to merge those two functions into this one and check price only once per collateral.

Reply:
changed ratioedDeposits to borrowingPower and merged in totalDeposits